Move session creation and notification sending

shutterbug2000 opened the pull request:
This helps reduce duplicate code and bugs that result from it.
Marked as a Draft for now since it needs testing with lots more games to ensure no regressions happen.

Tested with:
Minecraft: Wii U Edition
Tri-Force Heroes

Notable games to test still (these worked before this change):
Splatoon
Mario Kart 8
Mario Kart 7
Kid Icarus: Uprising

shutterbug2000 committed:
Move session creation and notification sending
This helps reduce duplicate code and bugs that result from it

DaniElectra commented on globals/matchmaking_globals.go:
```diff
@@ -14,3 +14,4 @@ type CommonMatchmakeSession struct {
 
 var Sessions map[uint32]*CommonMatchmakeSession
 var CurrentGatheringID = nex.NewCounter(0)
+var CurrentMatchmakingCallID uint32
```
This can be a counter

DaniElectra commented on globals/matchmaking_utils.go:
```diff
@@ -140,6 +142,33 @@ func RemoveClientFromAllSessions(client *nex.Client) {
+	Sessions[sessionIndex].GameMatchmakeSession.StartedTime = nex.NewDateTime(0)
+	Sessions[sessionIndex].GameMatchmakeSession.StartedTime.UTC()
+	Sessions[sessionIndex].GameMatchmakeSession.SessionKey = make([]byte, 32)
+    rand.Read(Sessions[sessionIndex].GameMatchmakeSession.SessionKey)
```
Use tabs, not spaces

DaniElectra commented on globals/matchmaking_utils.go:
```diff
@@ -140,6 +142,33 @@ func RemoveClientFromAllSessions(client *nex.Client) {
+// CreateSessionByMatchmakeSession creates a gathering from a MatchmakeSession
+func CreateSessionByMatchmakeSession(matchmakeSession *match_making_types.MatchmakeSession, searchMatchmakeSession *match_making_types.MatchmakeSession, hostPID uint32) (*CommonMatchmakeSession, error, uint32) {
+	sessionIndex := GetAvailableGatheringID()
+	// This should in theory be impossible, as there aren't enough PIDs creating sessions to fill the uint32 limit.
+	// If we ever get here, we must be not deleting sessions properly
```
This comment does not apply anymore, and can be removed

DaniElectra commented on globals/matchmaking_utils.go:
```diff
@@ -168,6 +197,34 @@ func FindSessionByMatchmakeSession(searchMatchmakeSession *match_making_types.Ma
+	Sessions[sessionIndex].GameMatchmakeSession.StartedTime = nex.NewDateTime(0)
+	Sessions[sessionIndex].GameMatchmakeSession.StartedTime.UTC()
+	Sessions[sessionIndex].GameMatchmakeSession.SessionKey = make([]byte, 32)
+    rand.Read(Sessions[sessionIndex].GameMatchmakeSession.SessionKey)
```
Use tabs, not spaces

DaniElectra commented on matchmake-extension/auto_matchmake_postpone.go:
```diff
@@ -90,57 +81,12 @@ func autoMatchmake_Postpone(err error, client *nex.Client, callID uint32, anyGat
 	responsePacket.SetDestination(0xAF)
 	responsePacket.SetType(nex.DataPacket)
-	responsePacket.SetPayload(rmcResponseBytes)
+	responsePacket.SetPayload(rmcResponseBytes)	
```
remove the tab at the end lmao

DaniElectra commented on matchmake-extension/auto_matchmake_with_param_postpone.go:
```diff
@@ -0,0 +1,131 @@
+			logger.Error(err.Error())
+			return errCode
+		}
+	}else{
```
```suggestion
	} else {
```

jonbarrow commented on globals/matchmaking_utils.go:
```diff
@@ -140,6 +142,33 @@ func RemoveClientFromAllSessions(client *nex.Client) {
 
+// CreateSessionByMatchmakeSession creates a gathering from a MatchmakeSession
+func CreateSessionByMatchmakeSession(matchmakeSession *match_making_types.MatchmakeSession, searchMatchmakeSession *match_making_types.MatchmakeSession, hostPID uint32) (*CommonMatchmakeSession, error, uint32) {
+	sessionIndex := GetAvailableGatheringID()
+	// This should in theory be impossible, as there aren't enough PIDs creating sessions to fill the uint32 limit.
+	// If we ever get here, we must be not deleting sessions properly
+	if sessionIndex == 0 {
+		return &CommonMatchmakeSession{}, fmt.Errorf("No gatherings available!"), nex.Errors.RendezVous.LimitExceeded
+	}
```
Not an "issue" per say, but since we're already here refactoring stuff. I've been thinking about this, maybe we should just reset the counter instead of erroring here? I think it's a pretty safe assumption that if the uint32 limit is hit then the oldest sessions are very much dead (hell maybe even the people who made them are too)

DaniElectra commented on matchmake-extension/auto_matchmake_with_param_postpone.go:
```diff
@@ -0,0 +1,131 @@
+		}
+
+		session := common_globals.CommonMatchmakeSession{
+			SearchMatchmakeSession: searchMatchmakeSession,
+			GameMatchmakeSession:   matchmakeSession,
+		}
+
+		session = &session
+		session.GameMatchmakeSession.Gathering.ID = sessionIndex
+		session.GameMatchmakeSession.Gathering.OwnerPID = client.PID()
+		session.GameMatchmakeSession.Gathering.HostPID = client.PID()
+
+		session.GameMatchmakeSession.StartedTime = nex.NewDateTime(0)
+		session.GameMatchmakeSession.StartedTime.UTC()
+		session.GameMatchmakeSession.SessionKey = make([]byte, 32)
+
+		session.GameMatchmakeSession.MatchmakeParam.Parameters["@SR"] = nex.NewVariant()
+		session.GameMatchmakeSession.MatchmakeParam.Parameters["@SR"].TypeID = 3
+		session.GameMatchmakeSession.MatchmakeParam.Parameters["@SR"].Bool = true
+	
+		session.GameMatchmakeSession.MatchmakeParam.Parameters["@GIR"] = nex.NewVariant()
+		session.GameMatchmakeSession.MatchmakeParam.Parameters["@GIR"].TypeID = 1
+		session.GameMatchmakeSession.MatchmakeParam.Parameters["@GIR"].Int64 = 3
+	
+		session.GameMatchmakeSession.MatchmakeParam.Parameters["@Lon"] = nex.NewVariant()
+		session.GameMatchmakeSession.MatchmakeParam.Parameters["@Lon"].TypeID = 2
+		session.GameMatchmakeSession.MatchmakeParam.Parameters["@Lon"].Float64 = 0
+	
+		session.GameMatchmakeSession.MatchmakeParam.Parameters["@Lat"] = nex.NewVariant()
+		session.GameMatchmakeSession.MatchmakeParam.Parameters["@Lat"].TypeID = 2
+		session.GameMatchmakeSession.MatchmakeParam.Parameters["@Lat"].Float64 = 0
+	
+		session.GameMatchmakeSession.MatchmakeParam.Parameters["@CC"] = nex.NewVariant()
+		session.GameMatchmakeSession.MatchmakeParam.Parameters["@CC"].TypeID = 4
+		session.GameMatchmakeSession.MatchmakeParam.Parameters["@CC"].Str = "US"
+	
+		session.GameMatchmakeSession.MatchmakeParam.Parameters["@DR"] = nex.NewVariant()
+		session.GameMatchmakeSession.MatchmakeParam.Parameters["@DR"].TypeID = 1
+		session.GameMatchmakeSession.MatchmakeParam.Parameters["@DR"].Int64 = 0
+	}*/
```
Why is this commented out? Can it be removed?

shutterbug2000 replied:
This was for Tri-Force Heroes, I kept it commented in case it was required, but it's not :p

jonbarrow commented on globals/matchmaking_utils.go:
```diff
@@ -140,6 +142,33 @@ func RemoveClientFromAllSessions(client *nex.Client) {
+		GameMatchmakeSession:   matchmakeSession,
+	}
+
+	Sessions[sessionIndex] = &session
+	Sessions[sessionIndex].GameMatchmakeSession.Gathering.ID = sessionIndex
+	Sessions[sessionIndex].GameMatchmakeSession.Gathering.OwnerPID = hostPID
+	Sessions[sessionIndex].GameMatchmakeSession.Gathering.HostPID = hostPID
+
+	Sessions[sessionIndex].GameMatchmakeSession.StartedTime = nex.NewDateTime(0)
+	Sessions[sessionIndex].GameMatchmakeSession.StartedTime.UTC()
+	Sessions[sessionIndex].GameMatchmakeSession.SessionKey = make([]byte, 32)
+    rand.Read(Sessions[sessionIndex].GameMatchmakeSession.SessionKey)
+
```
Possible micro-optimization here as well. Instead of doing many index lookups and then modifying the data, we could instead just set all the data on the `session` struct right away and only assign it to `Sessions` later. Unsure how much of a difference that would make but logically it should be faster?

jonbarrow commented on globals/matchmaking_utils.go:
```diff
@@ -140,6 +142,33 @@ func RemoveClientFromAllSessions(client *nex.Client) {
+	Sessions[sessionIndex].GameMatchmakeSession.SessionKey = make([]byte, 32)
+    rand.Read(Sessions[sessionIndex].GameMatchmakeSession.SessionKey)
+
+	return Sessions[sessionIndex], nil, 0
```
Now that we have `MutexMap` in `nex-go` I wonder if that would be safer here rather than a normal `map`

shutterbug2000 replied:
I tried to do this, but wasn't sure how MutexMap is meant to work. Might look into it in a bit, but wanted to get the other changes pushed first :p

jonbarrow replied:
`MutexMap` is used to create concurrency-safe maps. If we have multiple threads/go routines reading and writing to the map at the same time, Go will panic

`MutexMap` is a generic type which provides a wrapper around a `map` and uses mutex operations to ensure only one thread/go routine accesses the map at a time

Basic usage would be

```go
func main() {
	mutexMap := NewMutexMap[keyType, valueType]()

	mutexMap.Set(key, value)
	value, ok := mutexMap.Get(key)
	mutexMap.Delete(key)
}
```

DaniElectra replied:
The real map isn't exposed but we have to range through it, so we can't use that atm. Using `Each` is not an option, since on `FindClientSession` we may return early while iterating

DaniElectra commented on globals/matchmaking_utils.go:
```diff
@@ -292,6 +349,185 @@ func AddPlayersToSession(session *CommonMatchmakeSession, connectionIDs []uint32
+		server.Send(messagePacket)
+	}
+
+	//This appears to be correct. Tri-Force Heroes uses 3.9.0, and has issues if this is ran.
+	//Minecraft, however, requires this to be ran.
+	//TODO: Check other games both pre and post 3.10.0 and validate.
```
```suggestion
	// This appears to be correct. Tri-Force Heroes uses 3.9.0, and has issues if these notifications are sent
	// Minecraft, however, requires this to be sent
	// TODO: Check other games both pre and post 3.10.0 and validate
```

jonbarrow commented on globals/matchmaking_utils.go:
```diff
@@ -90,7 +91,8 @@ func RemoveClientFromAllSessions(client *nex.Client) {
 			rmcMessage := nex.NewRMCRequest()
 			rmcMessage.SetProtocolID(notifications.ProtocolID)
-			rmcMessage.SetCallID(0xffff0000)
+			rmcMessage.SetCallID(CurrentMatchmakingCallID)
+			CurrentMatchmakingCallID+=1
```
Can just use the `Counter` type from `nex-go` here instead of managing this yourself

DaniElectra commented on globals/matchmaking_utils.go:
```diff
@@ -277,7 +334,7 @@ func FindSessionsByMatchmakeSessionSearchCriterias(lstSearchCriteria []*match_ma
 // AddPlayersToSession updates the given sessions state to include the provided connection IDs
 // Returns a NEX error code if failed
-func AddPlayersToSession(session *CommonMatchmakeSession, connectionIDs []uint32) (error, uint32) {
+func AddPlayersToSession(session *CommonMatchmakeSession, connectionIDs []uint32, initiatingClient *nex.Client) (error, uint32) {
```
We also need to pass the join message now that we are handling notifications here

DaniElectra commented on globals/matchmaking_utils.go:
```diff
@@ -292,6 +349,185 @@ func AddPlayersToSession(session *CommonMatchmakeSession, connectionIDs []uint32
+
+			notificationRequestMessage.SetParameters(notificationStream.Bytes())
+			notificationRequestBytes := notificationRequestMessage.Bytes()
+			//fmt.Println(hex.EncodeToString(notificationRequestBytes))
```
```suggestion
```

shutterbug2000 committed:
Code review fixes

DaniElectra commented on globals/matchmaking_utils.go:
```diff
@@ -5,6 +5,7 @@ import (
 	"math"
 	"strconv"
 	"strings"
+	"math/rand" 
```
can we use `crypto/rand`?

DaniElectra replied:
Also, there's a space at the end

DaniElectra commented on globals/matchmaking_utils.go:
```diff
@@ -292,6 +347,181 @@ func AddPlayersToSession(session *CommonMatchmakeSession, connectionIDs []uint32
+		notificationRequestMessage.SetParameters(notificationStream.Bytes())
+		notificationRequestBytes := notificationRequestMessage.Bytes()
+
+		target := server.FindClientFromPID(uint32(session.GameMatchmakeSession.Gathering.OwnerPID))
```
We have to check if `target` is nil, like we are doing on the other notifications

DaniElectra commented on matchmake-extension/join_matchmake_session_with_param.go:
```diff
@@ -23,8 +21,7 @@ func joinMatchmakeSessionWithParam(err error, client *nex.Client, callID uint32,
 	// TODO - More checks here
-
-	err, errCode := common_globals.AddPlayersToSession(session, []uint32{client.ConnectionID()})
+	err, errCode := common_globals.AddPlayersToSession(session, []uint32{client.ConnectionID()}, client, "")
```
The join message exists on `joinMatchmakeSessionParam`

DaniElectra commented on matchmake-extension/create_matchmake_session_with_param.go:
```diff
@@ -22,45 +20,21 @@ func createMatchmakeSessionWithParam(err error, client *nex.Client, callID uint3
 	}
 
-	err, errCode := common_globals.AddPlayersToSession(common_globals.Sessions[sessionIndex], []uint32{client.ConnectionID()})
+	err, errCode = common_globals.AddPlayersToSession(session, []uint32{client.ConnectionID()}, client, "")
```
The join message exists inside createMatchmakeSessionParam

shutterbug2000 committed:
Code review fixes

shutterbug2000 committed:
Remove stray space

DaniElectra commented:
Works with Mario Kart 7 :+1:

DaniElectra commented on globals/matchmaking_globals.go:
```diff
@@ -14,3 +14,4 @@ type CommonMatchmakeSession struct {
 
 var Sessions map[uint32]*CommonMatchmakeSession
 var CurrentGatheringID = nex.NewCounter(0)
+var CurrentMatchmakingCallID = nex.NewCounter(0)
```
The `CurrentMatchmakingCallID` should also be used on `UnregisterGathering`, `EndParticipation` and other places that use notifications

shutterbug2000 committed:
Use CurrentMatchmakingCallID in other protocols

DaniElectra commented on matchmake-extension/auto_matchmake_with_param_postpone.go:
```diff
@@ -0,0 +1,85 @@
+
+	searchMatchmakeSession := matchmakeSession.Copy().(*match_making_types.MatchmakeSession)
+	commonMatchmakeExtensionProtocol.cleanupSearchMatchmakeSessionHandler(searchMatchmakeSession)
+	sessionIndex := common_globals.FindSessionByMatchmakeSession(searchMatchmakeSession)
```
`AutoMatchmakeWithParam_Postpone` has search criteria, so I would find the session using `FindSessionsByMatchmakeSessionSearchCriterias`

jonbarrow commented:
Confirmed working in Mario Kart 8 :+1:

shutterbug2000 committed:
Change AutoMatchmakeWithParam_Postpone to use SearchCriteria
This should allow for more precise matchmaking
Also changes SearchMatchmakeSession to GameMatchmakeSession since a SearchCriteria session won't have a searchMatchmakeSession

shutterbug2000 commented:
This has now been testing working in Splatoon

DaniElectra commented:
@jonbarrow Any objections with merging this?

jonbarrow commented:
Nope! Looks good to me :+1: I see it's still a draft, so once it's marked as ready it can be merged

shutterbug2000 commented:
> Nope! Looks good to me 👍 I see it's still a draft, so once it's marked as ready it can be merged

I've marked it as ready!
